test/e2e/scale.go

```diff
@@ -0,0 +1,213 @@
+			if err != nil {
+				t.Errorf("CreateLatestService() = %v", err)
+				return nil
+			}
+			// Record the time it took to create the service.
+			latencies.Add("time-to-create", start)
+			names.Route = serviceresourcenames.Route(svc)
+			names.Config = serviceresourcenames.Configuration(svc)
+
+			// Send it to our cleanup logic (below)
+			cleanupCh <- names
+
+			logger.Infof("Wait for %s to become ready.", names.Service)
+			var domain string
+			err = test.WaitForServiceState(clients.ServingClient, names.Service, func(s *v1alpha1.Service) (bool, error) {
+				if s.Status.Domain == "" {
+					return false, nil
+				}
+				domain = s.Status.Domain
+				return test.IsServiceReady(s)
+			}, "ServiceUpdatedWithDomain")
+			if err != nil {
+				t.Errorf("WaitForServiceState(w/ Domain) = %v", err)
+				return nil
+			}
+			// Record the time it took to become ready.
+			latencies.Add("time-to-ready", start)
+
+			_, err = pkgTest.WaitForEndpointState(
+				clients.KubeClient,
+				logger,
+				domain,
+				pkgTest.Retrying(pkgTest.EventuallyMatchesBody(helloWorldExpectedOutput), http.StatusNotFound),
+				"WaitForEndpointToServeText",
+				test.ServingFlags.ResolvableDomain)
+			if err != nil {
+				t.Errorf("WaitForEndpointState(expected text) = %v", err)
+				return nil
+			}
+			// Record the time it took to get back a 200 with the expected text.
+			latencies.Add("time-to-200", start)
+
+			// Start probing the domain until the test is complete.
+			pm.Spawn(domain)
+
+			logger.Infof("%s is ready.", names.Service)
+			return nil
+		})
+	}
+
+	// Wait for all of the service creations to complete (possibly in failure),
+	// and signal the done channel.
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		if err := wg.Wait(); err != nil {
+			t.Fatalf("hmm, this go routine never returns errors: %v", err)
+		}
+	}()
+
+	for {
+		// As services get created, add logic to clean them up.
+		// When all of the creations have finished, then stop all of the active probers
+		// and check our SLIs against our SLOs.
+		// All of this has to finish within the configured timeout.
+		select {
+		case names := <-cleanupCh:
+			logger.Infof("Added %v to cleanup routine.", names)
+			test.CleanupOnInterrupt(func() { TearDown(clients, names, logger) }, logger)
+			defer TearDown(clients, names, logger)
+
+		case <-doneCh:
+			// This ProberManager implementation waits for minProbes before actually stopping.
+			if err := pm.Stop(); err != nil {
+				t.Fatalf("Stop() = %v", err)
+			}
+			// Check each of the local SLOs
+			pm.Foreach(func(domain string, p test.Prober) {
+				if err := test.CheckSLO(localSLO, domain, p); err != nil {
+					t.Errorf("CheckSLO() = %v", err)
+				}
+			})
+			// Check the global SLO
+			if err := test.CheckSLO(globalSLO, "aggregate", pm); err != nil {
+				t.Errorf("CheckSLO() = %v", err)
+			}
+			return
+
+		case <-timeoutCh:
+			// If we don't do this first, then we'll see tons of 503s from the ongoing probes
```

Comment:
But we're not doing it first. We're doing it in the end?

Reply:
It's subtle, but `defer`s are stacked, so technically by `defer`ing immediately before `t.Fatalf` this defer will happen before the rest (which clean up services).

Reply:
Yeah, but the way it reads is: hey, we stop probes, then we fatal. Rather than, hey we defer, then we fatal, then all the other defferred calls are going to execute.